File: 0907-koko-eating-bananas/0907-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:

        left, right = 1, max(piles)
        ans = right

        while left <= right:

            mid = (left + right) // 2
            if self.canEat(piles, h, mid):
                ans = min(ans, mid)
                right = mid - 1
            else:
                left = mid + 1
        
        return ans
                
    

    def canEat(self, piles, h, pace):

        total_h = 0

        for pile in piles:
            total_h += ceil(pile / pace)
        
        if total_h <= h:
            return True
            
        

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int res = INT_MAX;
        int high = *max_element(piles.begin(), piles.end());
        int low = 1;
        
        while (low <= high) {
            int speed = low + ((high - low) / 2);
            double hours = 0;
            for (int i = 0; i < piles.size(); i++) {
                hours += ceil(((double)piles[i]) / speed);
            }
            if (hours <= h) {
                res = min(speed, res);
                high = speed - 1;
            } else low = speed + 1;
        }
        return res;

    }
};